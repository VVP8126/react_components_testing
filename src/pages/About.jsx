import Users from "../mocks/users/Users";

const About = () => {
    return (
        <div className="container" data-testid="about-page">
            <h1>USERS</h1>
            <h3>List:</h3>
            <Users />
        </div>
    );
}

export default About;
