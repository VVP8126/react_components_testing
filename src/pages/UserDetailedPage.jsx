import { useParams } from "react-router-dom";

const UserDetailedPage = () => {
    
    let {id} = useParams();

    return (<div data-testid="userPageBlock">PASSED: {id}</div>);
}
export default UserDetailedPage;
