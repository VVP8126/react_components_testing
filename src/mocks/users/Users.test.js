import { render, screen } from "@testing-library/react";
import mockAxios from "axios";
import { MemoryRouter } from "react-router-dom";
// import userEvent from '@testing-library/user-event';
import UserDetailedPage from "../../pages/UserDetailedPage";
import { Route, Routes } from "react-router-dom";
import About from "../../pages/About";

import { create } from "react-test-renderer";

// jest.mock('axios'); // - From the original example

describe('Mock test', () => { 
    let response;
    
    beforeEach(() => {
        response = {data: [
            {"id":1, "username": "Bret"},
            {"id":2, "username": "Antonette"},
            {"id":3, "username": "Samantha"}
        ]};
    });
    
    test('Test axios call ', async() => {
      /** 
        // Original example - doesn't work !!!
        render(<Users />);
        axios.get.mockReturnValue(response);     
        const users = await screen.findAllByTestId("userItem");
        expect(users.length).toBe(3);
        expect(axios.get).toBeCalledTimes(1);
        screen.debug();
      */

        // Found custom example
        mockAxios.get.mockImplementationOnce(() => Promise.resolve(response));
        expect(mockAxios.get).toHaveBeenCalledTimes(0);
        // OR SO: // UserService.loadUsers();
        render(
          <MemoryRouter initialEntries={["/about"]} >            
            <Routes>
              <Route path='about' element={<About />} />
              <Route path='about/:id' element={<UserDetailedPage />} />
            </Routes>
          </MemoryRouter>
        );
        expect(mockAxios.get).toHaveBeenCalledTimes(1);
        const linkElements = await screen.findAllByTestId("userLinkItem");
        expect(linkElements.length).toBe(3);
        
    });

    it("Links render", async () => {
      mockAxios.get.mockImplementationOnce(() => Promise.resolve(response));
      let renderer = create(
        <MemoryRouter initialEntries={["/about"]}>
          <Routes>
            <Route path='about' element={<About />} />
            <Route path='about/:id' element={<UserDetailedPage />} />
          </Routes>
        </MemoryRouter>
      );
      expect(renderer.toJSON()).toMatchSnapshot();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

});