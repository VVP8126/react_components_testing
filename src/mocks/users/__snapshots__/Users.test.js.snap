// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mock test Links render 1`] = `
<div
  className="container"
  data-testid="about-page"
>
  <h1>
    USERS
  </h1>
  <h3>
    List:
  </h3>
  <div />
</div>
`;

exports[`Mock test renders correctly 1`] = `
<div
  className="container"
  data-testid="about-page"
>
  <h1>
    USERS
  </h1>
  <h3>
    List:
  </h3>
  <div />
</div>
`;
